### PR TITLE
[SPARK-14092]  [SQL] move shouldStop() to end of while loop

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

This PR rollback some changes in #11274 , which introduced some performance regression when do a simple aggregation on parquet scan with one integer column.

Does not really understand how this change introduce this huge impact, maybe related show JIT compiler inline functions. (saw very different stats from profiling).
## How was this patch tested?

Manually run the parquet reader benchmark, before this change: 

```
Intel(R) Core(TM) i7-4558U CPU @ 2.80GHz
Int and String Scan:                Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------
SQL Parquet Vectorized                   2391 / 3107         43.9          22.8       1.0X
```

After this change

``````
Java HotSpot(TM) 64-Bit Server VM 1.7.0_60-b19 on Mac OS X 10.9.5
Intel(R) Core(TM) i7-4558U CPU @ 2.80GHz
Int and String Scan:                Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------
SQL Parquet Vectorized                   2032 / 2626         51.6          19.4       1.0X```

``````

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala [hunk 1]
```diff
@@ -255,11 +255,11 @@ private[sql] case class DataSourceScan(
       |       $numOutputRows.add(numRows);
       |     }
       |
-      |     while (!shouldStop() && $idx < numRows) {
+      |     while ($idx < numRows) {
       |       int $rowidx = $idx++;
       |       ${consume(ctx, columns1).trim}
+      |       if (shouldStop()) return;
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: Can we add some comment somewhere to explain why shouldStop needs to be here? It'd be great to reference the JIRA ticket.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: It's here in the beginning.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: where is it?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: btw i'm not sure but i suspect this has to do with loop unrolling. jit stops unrolling the loop when shouldStop is part of the terminal condition.

Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala]: Can we add a comment around line 248 saying this loop is very perf sensitive and changes to it should be measured carefully?

**Comment by nongli** [comment]
LGTM

**Comment by davies** [comment]
Added comment, merging this into master.

**Pull request closed without merging**